File: Behavioral/template-method/examples/name-generator/Python/NameGenerator/generators/GermanNameGenerator.py
from generators.BaseNameGenerator import BaseNameGenerator, Gender
import random


class GermanNameGenerator(BaseNameGenerator):
    __first_male_names = ["Hans", "Heinz", "Horst", "Andreas"]
    __first_female_names = ["Anna", "Elke", "Birgit", "Stefanie"]
    __surnames = ["Müller", "Schröder", "Hartmann", "Schulze"]

    def generate_first_name(self, gender: Gender) -> str:
        if gender == Gender.MALE:
            return self.__first_male_names[random.randint(0, len(self.__first_male_names)-1)]
        return self.__first_female_names[random.randint(0, len(self.__first_female_names)-1)]

    def generate_surname(self, gender: Gender) -> str:
        return self.__surnames[random.randint(0, len(self.__surnames)-1)]

File: Behavioral/template-method/examples/name-generator/C#/NameGenerator/NameGenerator/Generators/GermanNameGenerator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NameGenerator.Generators
{
    public class GermanNameGenerator : BaseNameGenerator
    {
        private readonly string[] _firstMaleNames = new string[] { "Hans", "Heinz", "Horst", "Andreas" };
        private readonly string[] _firstFemaleNames = new string[] { "Anna", "Elke", "Birgit", "Stefanie" };
        private readonly string[] _secondNames = new string[] { "Müller", "Schröder", "Hartmann", "Schulze" };

        protected override string GenerateFirstName(Gender gender)
        {
            switch (gender)
            {
                case Gender.Male:
                    return _firstMaleNames[_random.Next(0, _firstMaleNames.Length)];
                case Gender.Female:
                    return _firstFemaleNames[_random.Next(0, _firstFemaleNames.Length)];
                default:
                    throw new ArgumentOutOfRangeException();
            }

        }

        protected override string GenerateSurname(Gender gender)
        {
            return _secondNames[_random.Next(0, _secondNames.Length)];
        }
    }
}
